### DISCUSSION TITLE
Neovide/Kitty-like Cursor Trail/Animation?

### DISCUSSION THREAD

**Discussion by pbnj** [ghostty-org/ghostty#4199] [Feature Requests, Ideas] [answered]
[neovide](https://neovide.dev/configuration.html#animation-trail-size) & [kitty](https://sw.kovidgoyal.net/kitty/conf/#opt-kitty.cursor_trail) have a cursor animation feature, called "cursor trail" (see demo gif below).

This feature is helpful as it allows users/viewers to visual track cursor movements and reduces confusion from cursor jumps (e.g. jumping across multiple vim/tmux/terminal buffers/panes/windows/...etc)

Would love to see a feature/option like this added to ghostty.

---

Neovide demo gif:

![](https://neovide.dev/assets/ShortCursorTrailLength.gif)

**Comment by rhodes-b** [comment]
I thought there was already an issue for this but I cant find it

**Comment by jrop** [comment]
I spent some time looking into how this might be added. I don't know Zig, so while I might try to piece together a PR, I'm less confident in my ability to write idiomatic Zig code. I noted the following observations:

> The part the draws the terminal to the "Surface" is:
> - src/renderer/OpenGL.zig
> - (+other impls, too)
>
> There is a function, addCursor, that seems to add an appropriate sprite to be rendered.
> In there, terminal.Screen seems to hold the screen state, including the cursor position.
>
> Screen.zig:Cursor: add additional state to track the prior location, and compute the current (i.e., animated frame) sprite location.

Perhaps Mitchel can offer some guidance on what additional things may need to change for something like cursor-smear/trails to be supported. The current renderer seems to be very "cell"-based, so I am reluctant to start hacking away without understanding what assumptions in the architecture are built on top of this.

**Comment by jrop** [comment]
Prior conversations:

- https://github.com/ghostty-org/ghostty/issues/1934
- https://github.com/ghostty-org/ghostty/issues/1167
- https://github.com/ghostty-org/ghostty/discussions/2355 :point_left: this one seems like where this conversation should converge.

**Comment by noirbizarre** [comment]
I am truly interested in this feature because as it has been a game changer this year for me (in Kitty).

I have been amazed by this tweet showcasing a `zig` implementation: https://x.com/zack_overflow/status/1853494692169896361 
Repository (https://github.com/zackradisic/tether) is still empty but maybe `@zackradisic` would be interested in this? (quoted to avoid ping)

**Comment by blurskye** [comment]
no updates?

**Reply by 00-kat** [reply]
None yet; no one has had time to work on this as far as I know.  I'm fairly certain that animations such as this would require a pretty thorough rewrite of the rendering pipeline, and I don't remember if Qwerasd finished that or not (or perhaps I'm misremembering and they had never started working on it in the first place).

**Comment by sommerper** [comment]
I just discovered this in Kitty and now I can't live without it. It makes it so easy to follow the cursor around the screen. I'm really hoping this will be implemented at some point!

**Reply by blurskye** [reply]
yea, cant live without it, sadly in neovim, kitty is just laggy, probably because it was written in python, it is fine for couple of thousand of lines, after that, boom laggy aah exp, i switched back to neovide

**Comment by 00-kat** [comment]
If you do not have much more to say than “I really want this feature!”, please use the button over here instead of sending everyone subscribed an email:
![An image with the upvote button on the main post circled.](https://github.com/user-attachments/assets/762462b2-0ff3-4161-9c71-9077f261b02e)

**Comment by 00-kat** [comment]
Cross-referencing #6912 and #6901.

**Reply by 00-kat** [reply, marked as answer]
Looks like everyone forgot about this discussion, but it has been implemented in #7648.
